sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java

```diff
@@ -83,30 +90,41 @@ public class ChangeStreamRecordMapper {
   private static final String CHILD_PARTITIONS_COLUMN = "child_partitions";
   private static final String PARENT_PARTITION_TOKENS_COLUMN = "parent_partition_tokens";
   private static final String TOKEN_COLUMN = "token";
+  private final Dialect dialect;
+  private final JsonFormat.Printer printer;
+  private final JsonFormat.Parser parser;
 
-  ChangeStreamRecordMapper() {}
+  ChangeStreamRecordMapper(Dialect dialect) {
+    this.dialect = dialect;
+
+    this.printer =
```

Comment:
I don't think we will call these concurrently, but do you know if they are thread-safe?

Reply:
I don't really see any documentation explicitly specifying this class as thread-safe, unfortunately. However, like you said, it doesn't seem like these objects will be accessed concurrently?

Reply:
I think not, we create a new mapper in the [Setup](https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dofn/ReadChangeStreamPartitionDoFn.java#L186-L187) method, which is called per SDF, so we (the implementers) won't call it concurrently, but I don't know if the backend will (Windmill in the case of Dataflow). I would think not.

It would be good to test this under load to verify we don't get any errors